SDK: SC-258 Protobuf generation & interfaces, unit tests, CI

alexdelorenzo opened the pull request:
This PR is for SC-258 and it contains:
 - Protobuf interface generation
 - Export payload encoders
 - Unit test payload parsing and validation
 - Add CI validation for `sdk-schema`

CI code validation passes on my end :+1: 

[You can browse through this branch here](https://github.com/alexdelorenzo/console/tree/sc-258/python/packages/sdk-schema)

NOTE: I saw [the draft Max made](https://github.com/serverless/console/pull/334) that uses [Buf.build](https://buf.build) to automatically generate Python classes from protobuf interfaces, but found that it does not output importable Python modules. 

[It looks like Python support from Buf is experimental and not as well supported as Go and JavaScript are with Buf](https://docs.buf.build/bsr/remote-packages/overview#supported-languages), and to use the files it generated would mean writing scripts to convert them into modules, which can be pretty hacky.

The code Buf outputs for Python is also non-idiomatic, lacks type annotations, lacks `async` support which is necessary for FastAPI support along with other modern web frameworks, and [there are more issues with Buf's backend generator for Python documented here](https://github.com/danielgtaylor/python-betterproto#motivation).

For those reasons, I chose to generate Python code for the protobuf interfaces using [`betterproto`](https://github.com/danielgtaylor/python-betterproto). It addresses the issues I mentioned above, and generates idiomatic Python code with type annotations, has `async` support and the code it generates is generally more pleasant to work with for both developers and consumers of this library.

Code generation via `betterproto` can be automated for packaging via CI like Buf currently does for the Node code it generates.

alexdelorenzo committed:
Add initial scaffolding for `sdk`

alexdelorenzo committed:
Apply PR 334 patch

alexdelorenzo committed:
Add minimal `setup.py`

alexdelorenzo committed:
Implement `serverlessSdk` and its methods

alexdelorenzo committed:
Use explicit exports for module

alexdelorenzo committed:
Get name and version from package metadata

alexdelorenzo committed:
Handle case when env var isn't supplied

alexdelorenzo committed:
Fix typo

alexdelorenzo committed:
Add test scaffold

alexdelorenzo committed:
Add optional deps for running tests

alexdelorenzo committed:
Add first test and scaffolding

alexdelorenzo committed:
Add `pytest` fixture and tests

alexdelorenzo committed:
Add `Options` to public exports

alexdelorenzo committed:
Add more tests for `serverlessSdk`

alexdelorenzo committed:
Add `importlib_metadata` backport from Python 3.8 to Python 3.7 to deps

alexdelorenzo committed:
Get package metadata using backported library

alexdelorenzo committed:
Use inline tests with `pytest`, relative imports. All tests pass.

alexdelorenzo committed:
Add minimum version for `importlib_metadata`

alexdelorenzo committed:
Add minimum versions for `importlib_metadata` and `pytest`

alexdelorenzo committed:
Add minimum version to `wheel` dep

alexdelorenzo committed:
Add scaffolding

alexdelorenzo committed:
Add initial Python 3.7 job to CI

alexdelorenzo committed:
Test to ensure `serverlessSdk._initialize()` takes `options` argument

alexdelorenzo committed:
Use apt name

alexdelorenzo committed:
Add `steps` to CI

alexdelorenzo committed:
Add Python setup action to `steps`

alexdelorenzo committed:
Add caching of dependencies to CI

alexdelorenzo committed:
`setup-python` does not support PEP 621 `pyproject.toml` for dependency caching

See https://github.com/actions/setup-python/issues/529#issuecomment-1367029699

alexdelorenzo committed:
Add SDK and deps step

alexdelorenzo committed:
Add unit test validation

alexdelorenzo committed:
Format comments

alexdelorenzo committed:
Use canonical introspection methods

alexdelorenzo committed:
Remove unused imports

alexdelorenzo committed:
Add `black` to deps for code formatting a la Prettier, add CI job for `black`

alexdelorenzo committed:
Add `ruff` as linter to deps, use it in CI linting

alexdelorenzo committed:
Migrate code to `black` formatting

alexdelorenzo committed:
Point `setup-python` in the right direction

alexdelorenzo committed:
Add `backports` to deps

alexdelorenzo committed:
Implement `generate_id()` for `TraceSpan`, add aliases

alexdelorenzo committed:
Set default number of bytes for generated ids

alexdelorenzo committed:
Implement `is_valid_name()` and add dep

alexdelorenzo committed:
Add `TraceSpan` scaffold

alexdelorenzo committed:
Add Python versions to backported deps

alexdelorenzo committed:
Add error handling and validation for names

alexdelorenzo committed:
Handle startTime logic, name, input, output. Add type aliases

alexdelorenzo committed:
Move type aliases to core

alexdelorenzo committed:
Add unit tests for new modules

alexdelorenzo committed:
Add unit tests for new modules

alexdelorenzo committed:
Python 3.7 doesn't ship with `Final`

alexdelorenzo committed:
Use raw strings

alexdelorenzo committed:
Python 3.7 doesn't ship with subscriptable `Pattern`

alexdelorenzo committed:
Chain exceptions for richer errors

alexdelorenzo committed:
Keep code formatter happy

alexdelorenzo committed:
Import `annotations`

alexdelorenzo committed:
Simplify logic

alexdelorenzo committed:
Ensure param can be optional

alexdelorenzo committed:
Use delayed evaluation for annotations

alexdelorenzo committed:
Remove `Options` in favor of using Python's keyword args

alexdelorenzo committed:
No need to export `Options`

alexdelorenzo committed:
Remove unused imports

alexdelorenzo committed:
Formatting

alexdelorenzo committed:
Use `black` formatting

alexdelorenzo committed:
Add tests for `TraceSpan`

alexdelorenzo committed:
Add callable signature test helper

alexdelorenzo committed:
Use signature helper

alexdelorenzo committed:
Add rich tests, inspect method signatures

alexdelorenzo committed:
Fix imports, remove unused imports

alexdelorenzo committed:
Add `endTime`

alexdelorenzo committed:
Add tag parsing logic

alexdelorenzo committed:
Ensure that `endTime = 0` is handled

alexdelorenzo committed:
Test `sdk.createTraceSpan()`

alexdelorenzo committed:
Bring fixture into top-level module

alexdelorenzo committed:
Add `sdk.createTraceSpan()`

alexdelorenzo committed:
Keep type checker and formatter happy

alexdelorenzo committed:
Move metadata to `base`, only use top-level module for exports

alexdelorenzo committed:
Move `ServerlessSdk` to `sdk`, only use top-level module for exports

alexdelorenzo committed:
Keep formatter and linter happy

alexdelorenzo committed:
Use constant

alexdelorenzo committed:
Test `generate_id()` and `get_ensure_resource_name()`

alexdelorenzo committed:
`black` formatting

alexdelorenzo committed:
Define some custom SDK exceptions that derive from Python's standard exceptions.

This allows users to catch all SDK exceptions, as well as catching Python exceptions they might expect

alexdelorenzo committed:
Define some custom SDK exceptions that derive from Python's standard exceptions.

This allows users to catch all SDK exceptions, as well as catching Python exceptions they might expect

alexdelorenzo committed:
Ensure ability to set `output` correctly

alexdelorenzo committed:
Fix test comparison, `black` formatting

alexdelorenzo committed:
Test modules should reflect the names of the modules they test

alexdelorenzo committed:
Test modules should reflect the names of the modules they test. add `tags` tests

alexdelorenzo committed:
Use errors from JS SDK, formatting

alexdelorenzo committed:
Use errors from JS SDK, formatting

alexdelorenzo committed:
Keep `black` happy

alexdelorenzo committed:
Add tag validation, optimize imports

alexdelorenzo committed:
Add `Tags` implementation, tag validation, and hook it into `TraceSpan`

alexdelorenzo committed:
Leave `__setitem__()` upon setting value

alexdelorenzo committed:
Handle `start_time` errors, add new exceptions

alexdelorenzo committed:
Add `parentSpan` and `rootSpan` context var logic

alexdelorenzo committed:
Clear up logic, apply `black`

alexdelorenzo committed:
Clean up modules, imports, ensure private modules aren't exported, run `black`

alexdelorenzo committed:
Use type-checked model for protobuf intermediary object, add `.toProtobufObject()`, add new deps

alexdelorenzo committed:
Pass `tags` to `.createTraceSpan()`

alexdelorenzo committed:
Names should match modules

alexdelorenzo committed:
Handle imports from submodules

alexdelorenzo committed:
Ensure `traceId` property lookups end at the top of the hierarchy, fix annotation/global error, ensure optional attrs are defined at class definition

alexdelorenzo committed:
Use correct name

alexdelorenzo committed:
Test for all params, move fixtures back to test module, fix type error

alexdelorenzo committed:
Use valid name. Bytes can be non-ASCII

alexdelorenzo committed:
Ensure exceptions have sane messages

alexdelorenzo committed:
Ensure intermediary obj reflects protocol, test method

alexdelorenzo committed:
Add tests for `.close()` and `output` validation

alexdelorenzo committed:
Fix `startTime` logic, test TraceSpan creation from sdk object, formatting

alexdelorenzo committed:
Remove duplicate fixture

alexdelorenzo committed:
Remove regex literal tokens from string, test for more invalid non-real numbers, ensure handling situations where truthiness of values can be `False`-y despite the values being valid

alexdelorenzo committed:
Test tag name and value validity

alexdelorenzo committed:
Ensure key names are validated correctly

alexdelorenzo committed:
Add more tests and cases against `Tags`

alexdelorenzo committed:
Add more tests and cases against `Tags`

alexdelorenzo committed:
Add tests for parent span

alexdelorenzo committed:
Move globals to top of module

alexdelorenzo committed:
Handle `datetime`s

alexdelorenzo committed:
Handle `datetime`s and ISO date strings

alexdelorenzo committed:
Keep linter happy

alexdelorenzo committed:
Match test name with module name

alexdelorenzo committed:
Use parsed `__name__` from root module, not local module name

alexdelorenzo committed:
Use postponed evaluation of type annotations since we're targeting `>=3.7`

alexdelorenzo committed:
Ensure context vars get used correctly, fix test

alexdelorenzo committed:
Ensure global is used, not local var

alexdelorenzo committed:
Use type annotations where appropriate

alexdelorenzo committed:
Don't reset context var on `close()`

alexdelorenzo committed:
Ensure `update()` validates tags

alexdelorenzo committed:
Initial build script

alexdelorenzo committed:
Use Python naming conventions for attribute and method names from spec

alexdelorenzo committed:
Use Python conventions for names, name the default span for context variable, since ContextVar doesn't let you pass a named `default` keyword argument

alexdelorenzo committed:
Move `buf` Python plugins into Python specific template as per: https://docs.buf.build/generate/usage#multiple-templates

alexdelorenzo committed:
Export `TracePayload` and `RequestResponse`

alexdelorenzo committed:
Move generated classes to submodule for distribution

alexdelorenzo committed:
Add tests

alexdelorenzo committed:
Generate Python modules from protobuf schemas, add build and install tasks, add tests

alexdelorenzo committed:
Generate Python modules from protobuf schemas, add build and install tasks, add tests

alexdelorenzo committed:
Remove unit tests from `install` task

alexdelorenzo committed:
Add build dependencies

alexdelorenzo committed:
Add SDK schema validation to CI

alexdelorenzo committed:
Add typing backports for Python 3.7

alexdelorenzo committed:
Add payload parsing test

alexdelorenzo committed:
Add deps

alexdelorenzo committed:
Use editable install

alexdelorenzo committed:
Change validation order to check formatting, linting and testing in that order

alexdelorenzo committed:
Add trailing line

alexdelorenzo committed:
Set version to `0.0.0`

alexdelorenzo committed:
Only run CI unit tests if there are changes to SDK code

alexdelorenzo committed:
Only install module and dependencies if there are changes to SDK code

alexdelorenzo committed:
Only install and run unit tests upon schema changes, change order of validation to formatting, linting and testing

medikoo reviewed (commented):
@alexdelorenzo after merging #342 this needs to be rebased against `main`

Mmarzex reviewed (changes requested):
@alexdelorenzo That's fine if the remote plugin from buf doesn't work for Python. Can we do something similar to what we do for node then?

We use `ts-proto` to generate our node `sdk-schema` and buf just leverages the locally installed version to generate typescript classes.

https://github.com/serverless/console/blob/main/proto/buf.gen.yaml#L3

alexdelorenzo committed:
Fix after rebase

alexdelorenzo committed:
Remove trailing line

alexdelorenzo committed:
Run black

alexdelorenzo committed:
Ensure tests run

alexdelorenzo committed:
Get Buf working with `betterproto`

This was a pain, the trick was to install `betterproto` with the compiler plugin in `$PATH`. The plugin `name` has semantic meaning in Buf, it is the name of the plugin flags you'd pass to `protoc`.

alexdelorenzo committed:
Ensure we use the Python template. Using `buf.gen.yaml` would require installing NPM toolchain.

alexdelorenzo committed:
Remove Pythoin generation from main Buf setup

alexdelorenzo committed:
Add linting and conditions

medikoo commented on .github/workflows/python-validate.yml:
```diff
@@ -77,3 +77,53 @@ jobs:
+          cd python/packages/sdk-schema
+
+          python3 -m pip install black
+          python3 -m black --check .
```
As commented in other PR. It'll be probably better to have one `black` call for whole `python` directory

medikoo commented on .github/workflows/python-validate.yml:
```diff
@@ -77,3 +77,53 @@ jobs:
+          cd python/packages/sdk-schema
+
+          python3 -m pip install ruff
+          python3 -m ruff .
```
Similarly linting probably could work (be done against `python`)

medikoo commented on .github/workflows/python-validate.yml:
```diff
@@ -77,3 +77,53 @@ jobs:
+          python3 -m pip install ruff
+          python3 -m ruff .
+
+      - name: Run SDK Schema unit tests
```
Shouldn't those tests be run after we have proto changes linted and build?

alexdelorenzo replied:
Good point 👍

alexdelorenzo committed:
Run schema unit tests after proto changes are linted and built

alexdelorenzo committed:
Check code from the package level

alexdelorenzo committed:
Check formatting and linting dynamically

alexdelorenzo committed:
Pass `pyproject_fmt` metadata validation in CI

medikoo commented on .github/workflows/python-validate.yml:
```diff
@@ -56,19 +58,33 @@ jobs:
-          python3 -m ruff sdk
+          python3 -m ruff .
 
       - name: Run unit tests
```
Let's name this step `Run SDK unit tests`

medikoo reviewed (commented):
Thanks @alexdelorenzo looks good to me. I have just one minor not necessarily related to this changes suggestion.

I let  @Mmarzex approve before merging

alexdelorenzo committed:
Rename SDK unit test validation step

medikoo reviewed (approved):
Thank you @alexdelorenzo !
